instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py

```diff
@@ -12,117 +12,162 @@
+        if call_id:
+            tool_call_dict["id"] = call_id
+
+        tool_type = get_property_value(tool_call, "type")
+        if tool_type:
+            tool_call_dict["type"] = tool_type
+
+        func = get_property_value(tool_call, "function")
+        if func:
+            tool_call_dict["function"] = {}
+
+            name = get_property_value(func, "name")
+            if name:
+                tool_call_dict["function"]["name"] = name
+
+            arguments = get_property_value(func, "arguments")
+            if capture_content and arguments:
+                if isinstance(arguments, str):
+                    arguments = arguments.replace("\n", "")
+                tool_call_dict["function"]["arguments"] = arguments
 
-        if hasattr(tool_call, "function"):
-            tool_call_dict["function"] = {
-                "name": getattr(tool_call.function, "name", ""),
-                "arguments": getattr(tool_call.function, "arguments", ""),
-            }
         calls.append(tool_call_dict)
     return calls
 
 
-def set_event_prompt(span, prompt):
-    span.add_event(
-        name="gen_ai.content.prompt",
-        attributes={
-            GenAIAttributes.GEN_AI_PROMPT: prompt,
-        },
+def set_server_address_and_port(client_instance, attributes):
+    base_client = getattr(client_instance, "_client", None)
+    base_url = getattr(base_client, "base_url", None)
+    if not base_url:
+        return
+
+    port = -1
+    if isinstance(base_url, URL):
+        attributes[ServerAttributes.SERVER_ADDRESS] = base_url.host
+        port = base_url.port
+    elif isinstance(base_url, str):
+        url = urlparse(base_url)
+        attributes[ServerAttributes.SERVER_ADDRESS] = url.hostname
+        port = url.port
+
+    if port and port != 443 and port > 0:
+        attributes[ServerAttributes.SERVER_PORT] = port
+
+
+def get_property_value(obj, property_name):
+    if isinstance(obj, dict):
+        return obj.get(property_name, None)
+
+    return getattr(obj, property_name, None)
+
+
+def message_to_event(message, span_ctx, capture_content):
+    attributes = {
+        GenAIAttributes.GEN_AI_SYSTEM: GenAIAttributes.GenAiSystemValues.OPENAI.value
+    }
+    role = get_property_value(message, "role")
+    content = get_property_value(message, "content")
+
+    body = {}
+    if capture_content and content:
+        body["content"] = content
+    if role == "assistant":
+        tool_calls = extract_tool_calls(message, capture_content)
+        if tool_calls:
+            body = {"tool_calls": tool_calls}
+    elif role == "tool":
+        tool_call_id = get_property_value(message, "tool_call_id")
+        if tool_call_id:
+            body["id"] = tool_call_id
+
+    return Event(
+        name=f"gen_ai.{role}.message",
+        attributes=attributes,
+        body=body if body else None,
+        trace_id=span_ctx.trace_id,
+        span_id=span_ctx.span_id,
+        trace_flags=span_ctx.trace_flags,
```

Comment:
These should be set by default at emit() time, are you adding them explicitly because they do not match?

Reply:
great catch! Changed to implicit context, except the streaming case where the span associated with the stream is not current after the original method returns.